File: code/list/flat.py

#data = [["a"], ["b",["a"]], ["c",["b",["a"]]], ["d",["c",["b",["a"]]]],"e"]

lista=["a"]
listb=["b",lista]
listc=["c",listb]
listd=["d",listc]
data=[lista,listb,listc,listd,"e"]

print data

def flat(input,output):
    map(lambda x: flat(x,output) if type(x) is list else output.append(x),input)

def flatouput(input,output):
    for elem in input:
        if type(elem) is list:
            flatmap(elem,output)
        else:
            output.append(elem)
            
def flatmap(input):
    ret =[]
    for elem in input:
        ret += flatmap(elem) if type(elem) is list else elem
    return ret

print flatmap(data)

File: code/list/flat.go
package main

import (
        "fmt"

)

func flat(list []interface{}) []interface{} {
	var ret []interface{}

	for _,v := range list {
		switch t := v.(type) {
		default:
			ret = append(ret,t)
		case []interface{}:
			for _, e :=range flat(t) {
				ret = append(ret,e)
			}
		}

	}
	return ret
}

func main()  {

        lista := []interface{} {"a"}
	listb := []interface{} {"b",lista}
	listc := []interface{} {"c",listb}
	listd := []interface{} {"d",listc}
        data := []interface{} { lista, listb, listc, listd, "e"}

	fmt.Println(data)
	fmt.Println(flat(data))

}
